server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -119,6 +123,39 @@ public NodePath(Path path) throws IOException {
             }
         }
 
+        public void addShard(ShardId shardId) {
+            this.indices.computeIfAbsent(shardId.getIndexName(), k -> new ArrayList()).add(shardId);
+            this.numShards++;
+        }
+
+        public void removeShard(ShardId shardId) {
+            ArrayList<ShardId> shardIds = this.indices.get(shardId.getIndexName());
+            if (shardIds != null && shardIds.contains(shardId)) {
+                shardIds.remove(shardId);
+                this.numShards--;
+                if (shardIds.size() == 0) {
+                    this.indices.remove(shardId.getIndexName());
+                }
+            }
+        }
+
+        public void removeIndex(String index) {
+            ArrayList<ShardId> shardIds = this.indices.get(index);
+            if (shardIds != null) {
+                this.numShards -= shardIds.size();
+                this.indices.remove(index);
+            }
+        }
+
+        public int getNumShards(String index) {
+            ArrayList<ShardId> shardIds = this.indices.get(index);
+            return shardIds == null ? 0 : shardIds.size();
+        }
+
+        public int getNumShards() {
+            return numShards;
```

Comment:
This tracks allocated shards (i.e. for which there is an IndexService / IndexShard). I'm not sure  whether it's worth duplicating the state for IndicesService/IndexService about allocated shards here, as it introduces the risk of these two getting out of sync, and wrong book-keeping to happen here.

Reply:
I moved adding/removing shard operations from NodeEnv to IndexService, and add assert to make sure they always got synced in creating/deleting index or shard. 
State in NodeEnv is data path level, key is data path, whereas IndexService shards are index level.